packages/dht/src/dht/PeerManager.ts

```diff
@@ -0,0 +1,339 @@
+
+export interface PeerManagerConfig {
+    numberOfNodesPerKBucket: number
+    maxNeighborListSize: number
+    getClosestContactsLimit: number
+    ownPeerId: PeerID
+    connectionManager: ConnectionManager
+    nodeName: string
+    createDhtPeer: (peerDescriptor: PeerDescriptor) => DhtPeer
+}
+
+export interface PeerManagerEvents {
+    newContact: (peerDescriptor: PeerDescriptor, closestPeers: PeerDescriptor[]) => void
+    contactRemoved: (peerDescriptor: PeerDescriptor, closestPeers: PeerDescriptor[]) => void
+    newKbucketContact: (peerDescriptor: PeerDescriptor, closestPeers: PeerDescriptor[]) => void
+    kbucketContactRemoved: (peerDescriptor: PeerDescriptor) => void
+    newOpenInternetContact: (peerDescriptor: PeerDescriptor, closestPeers: PeerDescriptor[]) => void
+    openInternetContactRemoved: (peerDescriptor: PeerDescriptor, closestPeers: PeerDescriptor[]) => void
+    newRandomContact: (peerDescriptor: PeerDescriptor, closestPeers: PeerDescriptor[]) => void
+    randomContactRemoved: (peerDescriptor: PeerDescriptor, closestPeers: PeerDescriptor[]) => void
+    kBucketEmpty: () => void
+}
+
+interface IPeerManagerTest {
+    getNeighborList: () => SortedContactList<DhtPeer>
+    getKBucketSize: () => number
+    getKBucketPeers: () => PeerDescriptor[]
+}
+
+export class PeerManager extends EventEmitter<PeerManagerEvents> implements IPeerManager {
+
+    private bucket?: KBucket<DhtPeer>
+    private neighborList?: SortedContactList<DhtPeer>
+    private openInternetPeers?: SortedContactList<DhtPeer>
+    private randomPeers?: RandomContactList<DhtPeer>
+    public readonly connections: Map<PeerIDKey, DhtPeer> = new Map()
+    private readonly config: PeerManagerConfig
+    private stopped = false
+
+    private testInterface: IPeerManagerTest = {
+        getNeighborList: () => { return this.neighborList! },
+        getKBucketSize: () => { return this.bucket!.count() },
+        getKBucketPeers: () => { return this.bucket!.toArray().map((peer) => peer.getPeerDescriptor()) }
+    }
+    public testInterfaceType?: IPeerManagerTest 
+
+    constructor(config: PeerManagerConfig) {
+        super()
+        this.config = config
+        this.initKBuckets(config.ownPeerId)
+    }
+
+    private initKBuckets = (selfId: PeerID) => {
+        this.bucket = new KBucket<DhtPeer>({
+            localNodeId: selfId.value,
+            numberOfNodesPerKBucket: this.config.numberOfNodesPerKBucket,
+            numberOfNodesToPing: this.config.numberOfNodesPerKBucket
+        })
+
+        this.bucket.on('ping', (oldContacts: DhtPeer[], newContact: DhtPeer) => this.onKBucketPing(oldContacts, newContact))
+        this.bucket.on('removed', (contact: DhtPeer) => this.onKBucketRemoved(contact))
+        this.bucket.on('added', (contact: DhtPeer) => this.onKBucketAdded(contact))
+        this.bucket.on('updated', (_oldContact: DhtPeer, _newContact: DhtPeer) => {
+            // TODO: Update contact info to the connection manager and reconnect
```

Comment:
If there is a linear ticket, we could add link to that?

Reply:
@juslesan ?

Reply:
Made a ticket https://linear.app/streamr/issue/NET-1141/kbucket-on-updated